tachyon/crypto/commitments/fri/two_adic_fri_pcs.h

```diff
@@ -0,0 +1,403 @@
+// Copyright (c) 2022 The Plonky3 Authors
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.plonky3 and the LICENCE-APACHE.plonky3
+// file.
+
+#ifndef TACHYON_CRYPTO_COMMITMENTS_FRI_TWO_ADIC_FRI_PCS_H_
+#define TACHYON_CRYPTO_COMMITMENTS_FRI_TWO_ADIC_FRI_PCS_H_
+
+#include <algorithm>
+#include <map>
+#include <memory_resource>
+#include <optional>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+#include "absl/container/btree_map.h"
+#include "absl/container/flat_hash_map.h"
+
+#include "tachyon/base/bits.h"
+#include "tachyon/crypto/commitments/fri/two_adic_fri_proof.h"
+#include "tachyon/crypto/commitments/fri/two_adic_fri_prover.h"
+#include "tachyon/crypto/commitments/fri/two_adic_fri_verifier.h"
+#include "tachyon/crypto/commitments/merkle_tree/binary_merkle_tree/binary_merkle_tree.h"
+#include "tachyon/crypto/commitments/mixed_matrix_commitment_scheme.h"
+#include "tachyon/math/finite_fields/extension_field_traits_forward.h"
+#include "tachyon/math/geometry/dimensions.h"
+#include "tachyon/math/matrix/matrix_types.h"
+#include "tachyon/math/matrix/matrix_utils.h"
+#include "tachyon/math/polynomials/univariate/evaluations_utils.h"
+#include "tachyon/zk/air/plonky3/base/two_adic_multiplicative_coset.h"
+#include "tachyon/zk/air/plonky3/challenger/challenger.h"
+
+namespace tachyon::crypto {
+
+template <typename ExtF, typename InputMMCS, typename ChallengeMMCS,
+          typename Challenger>
+class TwoAdicFriPCS {
+ public:
+  using F = typename math::ExtensionFieldTraits<ExtF>::BaseField;
+  using Coset = zk::air::plonky3::TwoAdicMultiplicativeCoset<F>;
+  using Commitment = typename InputMMCS::Commitment;
+  using ProverData = typename InputMMCS::ProverData;
+  using Proof = typename InputMMCS::Proof;
+  using InputProof = std::vector<BatchOpening<InputMMCS>>;
+  using TwoAdicFriProof = crypto::TwoAdicFriProof<ChallengeMMCS, InputProof, F>;
+
+  using Points = std::vector<std::vector<ExtF>>;
+
+  using OpenedValuesForMat = std::vector<std::vector<ExtF>>;
+  using OpenedValuesForRound = std::vector<OpenedValuesForMat>;
+  using OpenedValues = std::vector<OpenedValuesForRound>;
+
+  TwoAdicFriPCS() = default;
+  TwoAdicFriPCS(InputMMCS&& mmcs, TwoAdicFriConfig<ChallengeMMCS>&& fri)
+      : mmcs_(std::move(mmcs)), fri_(std::move(fri)) {}
+
+  Coset GetNaturalDomainForDegree(size_t degree) {
+    size_t log_n = base::bits::CheckedLog2(degree);
+    return Coset(log_n, F::One());
+  }
+
+  [[nodiscard]] bool Commit(std::vector<Coset>& cosets,
+                            std::vector<math::RowMajorMatrix<F>>& matrices,
+                            Commitment* commitment, ProverData* prover_data) {
+    size_t coset_size = cosets.size();
+    std::vector<math::RowMajorMatrix<F>> ldes(coset_size);
```

Comment:
Let's just reserve here.

Reply:
I'll just change this to use `base::CreateVector`!